runtime/src/iree/hal/utils/event_semaphore.h

```diff
@@ -0,0 +1,57 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_HAL_UTILS_EVENT_SEMAPHORE_H_
+#define IREE_HAL_UTILS_EVENT_SEMAPHORE_H_
+
+#include <stdint.h>
+
+#include "iree/base/api.h"
+#include "iree/hal/utils/event_pool.h"
+#include "iree/hal/utils/pending_queue_actions.h"
+#include "iree/hal/utils/timepoint_pool.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// Creates an IREE HAL semaphore with the given |initial_value|.
+//
+// |symbols| and |symbol_user_data| must outlive the created semaphore.
+//
+// The HAL semaphore are backed by iree_event_t or CUevent/hipEvent_t objects
+// for different timepoints along the timeline under the hood. Those timepoints
+// will be allocated from the |timepoint_pool|.
+//
+// This semaphore is meant to be used together with a pending queue actions; it
+// may advance the given |pending_queue_actions| if new values are signaled.
+//
+// Thread-safe; multiple threads may signal/wait values on the same semaphore.
+iree_status_t iree_hal_event_semaphore_create(
```

Comment:
This is similar to `iree_hal_semaphore_create`... what makes it an "event semaphore"? Maybe just expand on the first line of this comment, but I'm also wondering if a new file is the right place for this.

Shoiuld this API even take `iree_hal_event_impl_symtable_t` and `iree_hal_timepoint_pool_t`? Maybe the API could be added to `iree/hal/event.h` instead, with the implementation pulling from existing struct members instead of needing these parameters here?

(Common theme here is "could these be generic across all HAL drivers?" The mentions of CUDA/HIP in many places are suspicious... what do they do differently than other drivers?)